Functions for the base direction feature

afs opened the pull request:
This addresses #154

#112 has been used for a different discussion.

---

The SPARQL grammar already has the base direction update #153.

This PR adds the functions `LANGDIR`, `STRLANGDIR`, `hasLANGDIR` and `hasLANG`.

It includes the non-grammar content from #113 with comments incorporated.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/pull/172.html" title="Last updated on Dec 20, 2024, 9:19 AM UTC (262dee4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/172/5dba9fd...262dee4.html" title="Last updated on Dec 20, 2024, 9:19 AM UTC (262dee4)">Diff</a>

afs commented:
Tests: https://github.com/w3c/rdf-tests/pull/156

Tpt commented on spec/index.html:
```diff
@@ -6238,16 +6238,203 @@ <h5>LANG</h5>
+              <p>
+                The datatype of a literal with a
+                <a data-cite="RDF12-CONCEPTS#dfn-language-tag">language tag</a> 
+                has a datatype of <code>rdf:langString</code>.
```
The sentence already starts with "The datatype  of":
```suggestion
                is <code>rdf:langString</code>.
```

Tpt commented on spec/index.html:
```diff
@@ -6238,16 +6238,203 @@ <h5>LANG</h5>
+                <a data-cite="RDF12-CONCEPTS#dfn-language-tag">language tag</a> 
+                and a
+                <a data-cite="RDF12-CONCEPTS#dfn-base-direction">base direction</a>
+                has a datatype of <code>rdf:dirLangString</code>.
```
Same:
```suggestion
                is <code>rdf:dirLangString</code>.
```

afs replied:
Done.

Tpt commented on spec/index.html:
```diff
@@ -6355,23 +6553,106 @@ <h5>STRDT</h5>
+              as specified by the arguments.
+            </p>
+            <p>
+              The argument `langTag` must not be an empty string.
```
Should we state it must (or should?) be a valid [language tag](https://www.w3.org/TR/rdf11-concepts/#dfn-language-tag)

afs replied:
I think "should" is better because RDF does not we don't insist on valid language tags - and also "valid" is a potentially moving target.

Change made and also in `STRDIRLANG`.

Also - using MUST and SHOULD.

Tpt commented on spec/index.html:
```diff
@@ -6431,9 +6712,27 @@ <h6>Argument Compatibility Rules</h6>
+                <li>
+                  The first argument is a literal with datatype <code>rdf:dirLangString</code>,
+                  and the second argument has datatype <code>xsd:string</code>
+                </li>
```
Genuine question: why don't we state that `"abc"@en--ltr` is compatible with `"abc"@en`?

afs replied:
`"abc"@en` could be rtl text. For some languages, there are multiple scripts, some of which are implicitly ltr and some are rtl.

In other words, it gets complicated (!) and the 1.1 argument compatibility rules didn't allow partial language tag matching.

Tpt replied:
Thank you! Make sense

Tpt commented on spec/index.html:
```diff
@@ -6498,7 +6816,8 @@ <h5>STRLEN</h5>
               function and returns an
-              <code>xsd:integer</code> equal to the length in characters of the lexical form of the
+              <code>xsd:integer</code> equal to the length in characters of the 
+              <a data-cite="RDF12-CONCEPTS#dfn-language-tag">language tag</a> of the
```
Should be lexical form

hartig replied:
```suggestion
              <a data-cite="RDF12-CONCEPTS#dfn-lexical-form">lexical form</a> of the
```

Tpt commented on spec/index.html:
```diff
@@ -6528,8 +6847,10 @@ <h5>SUBSTR</h5>
-              as the <code>source</code> input parameter but with a lexical form derived from
-              the substring of the lexical form of the source.</p>
+              as the <code>source</code> input parameter but with a 
+              <a data-cite="RDF12-CONCEPTS#dfn-language-tag">language tag</a> 
```
should be lexical form

hartig replied:
```suggestion
              <a data-cite="RDF12-CONCEPTS#dfn-lexical-form">lexical form</a> 
```

hartig commented on spec/index.html:
```diff
@@ -6355,23 +6553,106 @@ <h5>STRDT</h5>
+            <h5>STRLANGDIR</h5>
+            <pre class="prototype nohighlight"><span class="return">literal</span>  <span class="operator">STRLANGDIR</span>(<span class="type">xsd:string</span> lexicalForm, <span class="type">xsd:string</span> langTag, <span class="type">xsd:string</span> baseDirection)</pre>
+            <p>
+              The <code>STRLANG</code> function constructs a literal with 
```
```suggestion
              The <code>STRLANGDIR</code> function constructs a literal with 
```

afs commented on spec/index.html:
```diff
@@ -6238,16 +6238,203 @@ <h5>LANG</h5>
+              the function <code>LANGDIR</code> returns the
+              <a data-cite="RDF12-CONCEPTS#dfn-base-direction">base direction</a>
+              of the literal with datatype <code>rdf:dirLangString</code>.
+              Otherwise, the function returns the empty string.
```
First part done.

"If - Otherwise," is used in many other places and I'd like to keep that style.
"For all other"  is not an appropriate wording in every other place.

afs commented:
I hope I've addressed all comments.

The check ("Verift common files are consistent") is failing because the "participants" PR has not been merged.

afs commented:
Rebased for changes to common files.
